### ISSUE TITLE
RC16 tarball (aka the one shipping Guzzle v7) requires PHP >=8.1

### ISSUE THREAD

**Issue by ddavness** [mstilkerich/rcmcarddav#386]
The plugin tarball for Roundcube 1.6 ships a vendored version of composer that absolutely requires the system to run PHP 8.1:

```
Composed detected issues in your platform:

Your Composer dependencies require a PHP version ">= 8.1.0". You are running 7.4.28.
```

Is this intentional/necessary? Given that RC 1.6 is supposed to still support PHP 7.3 and above.

**Comment by mstilkerich** [comment]
Nope, it’s not intended. I guess it‘s just a result of the having PHP 8.1 in the system used to build the tarball… I‘ll have to read up the composer docs to figure an option on how to avoid this.

**Comment by mstilkerich** [comment]
Hello,

 I created v4.4.1, which is just a rebuild of the tarballs instructing composer to mimic a PHP 7.1.0 platform (this is the minimum needed by RCMCardDAV). So composer should not select any packages requiring a higher PHP version when fetching the dependencies. I hope this works now, at least `composer check-platform-reqs` now succeeds on an Ubuntu 20.04 (PHP 7.4) system.

**Comment by ddavness** [comment]
Negative - when you ask composer to mimic a PHP 7.1.0 platform it will just revert back to Guzzle v6 - and now we have the same issue as #385.

Guzzle v7 requires at least PHP 7.2.5 - RC16 actually requires PHP 7.3 or higher - therefore I guess that the RC16 tarball should also require at least PHP 7.3

**Comment by mstilkerich** [comment]
Thanks for the feedback. PHP 7.1.0 is what RCMCardDAV requires, and therefore I think it's appropriate if it came with a version 6 Guzzle in the release tarball. The fundamental issue here is having conflicting versions of the same package in roundcube and RCMCardDAV, which is avoided when using composer to install RCMCardDAV.

I was thinking of another solution / workaround: Instead of prepending the autoloader of RCMCardDAV tarball as composer does by default, append it to the list of autoloaders. By that, the roundcube/vendor directory will always be checked first. This avoids the issue we currently have that some Guzzle classes have already been loaded from the roundcube vendor directory before the plugin was loaded by roundcube, and subsequently after the plugin is loaded additional Guzzle classes are loaded from the carddav/vendor directory. I expect there could even be incompatibilities between patch releases of Guzzle, as a pure change of internal interfaces would not require an increase of the major version of Guzzle. In the end, it is basically impossible to ship tarballs matching the exact versions of the dependencies that already come with the various roundcube versions, therefore I think this would be a good approach to avoid this situation. So my proposal would be:

- Ship only one RCMCardDAV release tarball coming with dependencies that work with the minimum RCMCardDAV requirements (i. e. Guzzle 6).
- When using a roundcube version < 1.5 where no Guzzle is included with roundcube, the version included in the tarball would be used.
- When using a roundcube version, the version coming with roundcube will be used. This means for roundcube 1.5 Guzzle 6 would be used, for 1.6 Guzzle 7 would be used.

I have updated the tarballs for 4.4.1 with this scheme and tried the 1.6 tarball with roundcube 1.6, it appears to work in order.

What do you think?

**Comment by mstilkerich** [comment]
btw for your use case (from the issue you linked) why would you not use composer to install rcmcarddav to avoid this mess?

**Comment by ddavness** [comment]
It's a fork, and the [upstream project](https://github.com/mail-in-a-box/mailinabox) (which is still being maintained) went for the tarball route.

I chose to keep most of the setup process as-is to avoid ugly merge conflicts when they push changes of their own.

**Comment by ddavness** [comment]
About the proposed solution - from an user standpoint, if it works, it works, and if one only needs one tarball, even better.

Unfortunately I've tested both patched tarballs on Roundcube 1.6-rc (tested just in case they were swapped around for some reason) and seemingly both tried to use Guzzle v6 (therefore, both errored out).

I'm trying to figure out if it's something on my end - did you by any chance use any specific configuration?

**Comment by mstilkerich** [comment]
Thanks for the feedback. I have now published this as v4.4.2 (because I needed to update the INSTALL documentation back to use one tarball). There is a risk for incompatibilities open, but I think this is the best I can do for something that is broken by design (mailinabox should really use composer to install). I have checked the dependencies coming with RCMCardDAV 4.4.2 and the current point releases of roundcube 1.4 - 1.6-rc and believe that there will not be any issues.
